File: dss/delivery/Line.py
# Copyright (C) 2010 Richard Lincoln
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA, USA

from numpy import pi, sqrt
from scipy.sparse import csr_matrix as sparse

from dss.delivery.PowerDeliveryElement import PowerDeliveryElement

# 5 kVAr of capacitive reactance at 345 kV to avoid open line problem.
CAP_EPSILON = complex(0.0, 4.2e-8)

class Line(PowerDeliveryElement):
    """A Line defaults to 3-phases and some typical symmetrical component
    data.

    Line impedances are specified in per unit length and are multiplied
    by the length when the primitive Y matrix is computed.  You may specify the
    impedances of the line either by symmetrical components or by R, X, and
    nodal C matrices  (also per unit length).  All C's is entered in nano
    farads.  The ultimate values are in the matrices.  If you specify matrices,
    then the symmetrical component values are ignored.  However, if you change
    any of the symmetrical component values the matrices will be recomputed.
    It is assumed you want to use symmetrical component values.  Don't mix data
    entry by matrix and by symmetrical components.  Note that if you change the
    number of phases, the matrices are reallocated and reinitialized with
    whatever is currently in the symmetrical component data.   Multi-phase,
    two-port line or cable.  Pi model.  Power delivery element described by its
    impedance.  Impedances may be specified by symmetrical component values or
    by matrix values.  Alternatively, you may simply refer to an existing
    LineCode object from which the impedance values will be copied.  Then you
    need only specify the length.  You can define the line impedance at a base
    frequency directly in a Line object definition or you can import the
    impedance definition from a LineCode object. Both of these definitions of
    impedance are quite similar except that the LineCode object can perform
    Kron reduction.  If the geometry property is specified all previous
    definitions are ignored. The DSS will compute the impedance matrices from
    the specified geometry each time the frequency changes.  Whichever
    definition is the most recent applies, as with nearly all DSS functions.
    Note the units property; you can declare any length measurement in whatever
    units you please.  Internally, everything is converted to meters. Just be
    sure to declare the units. Otherwise, they are assumed to be compatible
    with other data or irrelevant.
    """

    def __init__(self, bus1='', bus2='', lineCode='', length=0.0, r1=0.0,
            x1=0.0, r0=0.0, x0=0.0, c1=0.0, c0=0.0, switch=False, rg=0.0,
            xg=0.0, rho=0.0, geometry='', units="none", rMatrix=None,
            xMatrix=None, cMatrix=None, spacing=None, wires=None,
            *args, **kw_args):
        """Initialises a new 'Line' instance.
        """
        #: Name of bus for terminal 1. Node order definitions optional.
        self.bus1 = bus1

        #: Name of bus for terminal 2.
        self.bus2 = bus2

        #: Name of linecode object describing line impedances.  If you use a
        #  line code, you do not need to specify the impedances here.  The line
        #  code must have been PREVIOUSLY defined.  The values specified last
        #  will prevail over those specified earlier (left-to-right sequence of
        #  properties).  If no line code or impedance data are specified, line
        #  object defaults to 336 MCM ACSR on 4 ft spacing.
        self.lineCode = lineCode

        #: Length of line. If units do not match the impedance data, specify
        #  'units' property.
        self.length = length

        #: Positive-sequence Resistance, ohms per unit length.
        self.r1 = r1

        #: Positive-sequence Reactance, ohms per unit length.
        self.x1 = x1

        #: Zero-sequence resistance, ohms per unit length.
        self.r0 = r0

        #: Zero-sequence Reactance, ohms per unit length.
        self.x0 = x0

        #: Positive-sequence capacitance, nF per unit length.
        self.c1 = c1

        #: Zero-sequence capacitance, nF per unit length.
        self.c0 = c0

        #: Designates this line as a switch for graphics and algorithmic
        #  purposes.  SIDE EFFECT: Sets R1=0.001 X1=0.0. You must reset if you
        #  want something different.
        self.switch = switch

        #: Carson earth return resistance per unit length used to compute
        #  impedance values at base frequency.  For making better frequency
        #  adjustments.
        self.rg = rg

        #: Carson earth return reactance per unit length used to compute
        #  impedance values at base frequency.  For making better frequency
        #  adjustments.
        self.xg = xg

        #: Earth resistivity used to compute earth correction factor. Overrides
        #  Line geometry definition if specified.
        self.rho = rho

        #: Geometry code for LineGeometry Object. Supercedes any previous
        #  definition of line impedance. Line constants are computed for each
        #  frequency change or rho change. CAUTION: may alter number of phases.
        self.geometry = geometry

        #: Default is None - assumes length units match impedance units. Values
        #  are: "none", "mi", "km", "kft", "m", "me", "ft", "in", "cm"
        self.units = units

        #: Resistance matrix, lower triangle, ohms per unit length. Order of
        #  the matrix is the number of phases. May be used to specify the
        #  impedance of any line configuration.  For balanced line models, you
        #  may use the standard symmetrical component data definition instead.
        self.rMatrix = rMatrix

        #: Reactance matrix, lower triangle, ohms per unit length. Order of the
        #  matrix is the number of phases. May be used to specify the impedance
        #  of any line configuration.  For balanced line models, you may use
        #  the standard symmetrical component data definition instead.
        self.xMatrix = xMatrix

        #: Nodal Capacitance matrix, lower triangle, nf per unit length.  Order
        #  of the matrix is the number of phases.  May be used to specify the
        #  shunt capacitance of any line configuration.  For balanced line
        #  models, you may use the standard symmetrical component data
        #  definition instead.
        self.cMatrix = cMatrix

        self.spacing = spacing

        self.wires = [] if wires is None else wires

        super(Line, self).__init__(*args, **kw_args)


    def recalcElementData(self):

        n_phases = self.n_phases
        r1 = self.r1
        x1 = self.x1
        c1 = self.c1
        positive_sequence = self.parent.positive_sequence
        length = self.length

        z = sparse((n_phases, n_phases))
        z_inv = sparse((n_phases, n_phases))
        y_c = sparse((n_phases, n_phases))

        ONETHIRD = 1.0 / 3.0

        # Only time this is called is if symmetrical components are specified.
        z_tmp = complex(r1, x1) * 2.0

        # Handle special case for 1-phase line and/or positive sequence model.
        if n_phases == 1 or positive_sequence:
            # Long-line equivalent PI, but only for ckt_model='positive'.
            if positive_sequence and (c1 > 0.0):
                # Nominal PI parameters per unit length.
                z_s = complex(r1, x1)
                y_s = complex(0.0, 2*pi * base_frequency * c1)

                # Apply the long-line correction to obtain Zm and Ym.
                gamma_l = sqrt(mul(z_s, y_s))
                gamma_l = mul(gamma_l, length)

                exp_p = mul( complex(cos(gamma_l.imag()), sin(gamma_l.imag())),
                             exp(gamma_l.real()) )

                exp_2p = mul( complex(cos(0.5 * gamma_l.imag()),
                                      sin(0.5 * gamma_l.imag())),
                              exp(0.5 * gamma_l.real()) )

                exp_m  = numpy.inverse(exp_p)
                exp_2m = numpy.inverse(exp_2p)

                sinh_gl  = mul(exp_p - exp_m, 0.5)
                tanh2_gl = div(exp_2p - exp_2m, exp_2p + exp_2m)

                z_m = div( mul( mul(z_s, length), sinh_gl), gamma_l)
                y_m = div( mul( mul(y_s, length), tanh2_gl), mul(gamma_l, 0.5))

                # Rely on this function being called only once, unless R1, X1,
                # or C1 changes.
                r1 = z_m.real() / length
                x1 = z_m.imag() / length
                c1 = y_m.imag() / length / 2*pi / base_frequency

            # Zero sequence the same as positive sequence.
            r0 = r1
            x0 = x1
            c0 = c1

        z_s = mul( z_tmp + complex(r0, x0), ONETHIRD )
        z_m = mul( complex(r0, x0) - complex(r1, x1), ONETHIRD )

        y_c1 = 2*pi * base_frequency * c1
        y_c0 = 2*pi * base_frequency * c0

        y_s = mul(mul(complex(0.0, y_c1), 2.0) + complex(0.0, y_c0), ONETHIRD)
        y_m = mul( complex(0.0, y_c0) - complex(0.0, y_c1), ONETHIRD )

        for i in range(n_phases):
            z[i, i] = z_s
            y_c[i, i] = y_s

            for j in range(i):
                z_m[i, j] = z_m
                z_m[j, i] = z_m
                y_c[i, j] = y_m
                y_c[j, i] = y_m

        self.sym_components_changed = False


    def calcYPrim(self):
        """Returns the primitive Y (admittance) matrix for this element alone.
        """
        freq_mult = 1.0
        len_mult = 1.0

        if self.sym_components_changed:
            self.recalc_element_data()

        self._clear_y_prim()

        # Build series y_prim and Z matrix.
        if self.geometry_specified:
            raise NotImplementedError
        else:
            # Z is from line code or specified in line data.
            # Convert to per unit length.
            len_mult = length / conversion_factor # FUnitsConvert
            y_prim_freq = self.parent.solution.frequency
            freq_mult = y_prim_freq / base_frequency

            # Put in series RL.
            z_values = z.get_values_array_ptr()
            z_inv_values = z.get_values_array_ptr()

            # Correct the impedances for length and frequency.
            # Rg increases with frequency.
            # Xg modified by ln of sqrt(1/f).
            if xg != 0.0:
                xg_mod = 0.5 * k_xg * log(freq_mult)
            else:
                xg_mod = 0.0

            for i in range(n_order * n_order):
                z_inv_values[i] = complex(
                    z_values[i].real() + rg * (freq_mult - 1.0) * len_mult,
                    (z_values[i].imag() - xg_mod) * len_mult * freq_mult
                )

            numpy.invert(z_inv)

            for i in range(n_phases):
                for j in range(n_phases):
                    value = z_inv[i, j]
                    y_prim_series[i, j] = value
                    y_prim_series[i + n_phases, j + n_phases] = value
                    y_prim_series[i, j + n_phases] = -value
                    y_prim_series[j + n_phases, i] = -value

            y_prim = matrix(y_prim_series)

            for i in range(y_order):
                y_prim_series[i, i] = CAP_EPSILON

        # Now build the shunt admittances and add into y_prim.
        # Put half the Shunt Capacitive Admittance at each end.
        y_values = yc.get_values_array_ptr(n_order)

        if self.geometry_specified:
            # Values are already compensated for length and frequency.
            raise NotImplementedError

        else:
            # Regular line model - values computed per unit length at base
            # frequency.
            k = 0
            for i in range(n_phases):
                for j in range(n_phases):
                    k += 1 # Assume matrix in col order (1,1  2,1  3,1 ...).
                    value = complex(
                        0.0,
                        y_values[k].imag() * len_mult * freq_mult / 2.0
                    )
                    y_prim_shunt[i, j] = value
                    y_prim_shunt[i + n_phases, j + n_phases] = value

        # TODO: Account for open conductors.

        y_prim += y_prim_shunt

        self.y_prim_invalid = False


    def _clearYPrim(self):
        """Initialises the primitive Y matrix.
        """
        if self.y_prim_invalid:
            y_prim_series = sparse((y_order, y_order))
            y_prim_shunt  = sparse((y_order, y_order))
            y_prim        = sparse((y_order, y_order))

        else:
            y_prim_series[:] = 0.0
            y_prim_shunt[:]  = 0.0
            y_prim[:]        = 0.0

File: dss/src/main/java/com/ncond/dss/delivery/Line.java
/**
 * Copyright (C) 2008-2012 Electric Power Research Institute, Inc.
 * Copyright (C) 2009-2012 Richard Lincoln
 * All rights reserved.
 */
package com.ncond.dss.delivery;

import org.apache.commons.math.complex.Complex;

import com.ncond.dss.common.DSS;
import com.ncond.dss.common.DSSClassDefs;
import com.ncond.dss.general.LineCode;
import com.ncond.dss.general.LineGeometry;
import com.ncond.dss.parser.Parser;
import com.ncond.dss.shared.CMatrix;
import com.ncond.dss.shared.CommandList;
import com.ncond.dss.shared.LineUnits;

import static com.ncond.dss.common.Util.interpretYesNo;
import static com.ncond.dss.common.Util.interpretEarthModel;

import static java.lang.Math.sqrt;
import static java.lang.Math.log;


public class Line extends PDClass {

	public static final int NumPropsThisClass = 23;

	// 5 kvar of capacitive reactance at 345 kV to avoid open line problem;
	public static final Complex CAP_EPSILON = new Complex(0.0, 4.2e-8);

	public static LineObj activeLineObj;

	public static LineGeometry lineGeometryClass;
	public static LineCode lineCodeClass;

	public Line() {
		super();
		className = "Line";
		classType = classType + DSSClassDefs.LINE_ELEMENT;  // in both PD element list and line section lists

		activeElement = -1;
		lineCodeClass = null;
		lineGeometryClass = null;

		defineProperties();

		String[] commands = new String[numProperties];
		System.arraycopy(propertyName, 0, commands, 0, numProperties);
		commandList = new CommandList(commands);
		commandList.setAbbrevAllowed(true);
	}

	@Override
	protected void defineProperties() {
		numProperties = Line.NumPropsThisClass;
		countProperties();  // get inherited property count

		allocatePropertyArrays();

		// define property names
		propertyName[0] = "bus1";
		propertyName[1] = "bus2";
		propertyName[2] = "linecode";
		propertyName[3] = "length";
		propertyName[4] = "phases";
		propertyName[5] = "r1";
		propertyName[6] = "x1";
		propertyName[7] = "r0";
		propertyName[8] = "x0";
		propertyName[9] = "c1";
		propertyName[10] = "c0";
		propertyName[11] = "rmatrix";
		propertyName[12] = "xmatrix";
		propertyName[13] = "cmatrix";
		propertyName[14] = "Switch";
		propertyName[15] = "Rg";
		propertyName[16] = "Xg";
		propertyName[17] = "rho";
		propertyName[18] = "geometry";
		propertyName[19] = "units";
		propertyName[20] = "spacing";
		propertyName[21] = "wires";
		propertyName[22] = "earthModel";
		propertyName[23] = "cncables";
		propertyName[24] = "tscables";

		// define property help values
		propertyHelp[0] = "Name of bus to which first terminal is connected."+ DSS.CRLF+
				"Example:"+DSS.CRLF+
				"bus1=busname   (assumes all terminals connected in normal phase order)"+DSS.CRLF+
				"bus1=busname.3.1.2.0 (specify terminal to node connections explicitly)";
		propertyHelp[1] = "Name of bus to which 2nd terminal is connected.";
		propertyHelp[2] = "Name of linecode object describing line impedances."+DSS.CRLF+
				"If you use a line code, you do not need to specify the impedances here. "+
				"The line code must have been PREVIOUSLY defined. " +
				"The values specified last will prevail over those specified earlier (left-to-right " +
				"sequence of properties).  You can subsequently change the number of phases if symmetrical component quantities are specified." +
				"If no line code or impedance data are specified, the line object "+
				"defaults to 336 MCM ACSR on 4 ft spacing.";
		propertyHelp[3] = "Length of line. Default is 1.0. If units do not match the impedance data, specify \"units\" property. ";
		propertyHelp[4] = "Number of phases, this line.";
		propertyHelp[5] = "Positive-sequence Resistance, ohms per unit length. Setting any of R1, R0, X1, X0, C1, C0 forces " +
				"the program to use the symmetrical component line definition. See also Rmatrix.";
		propertyHelp[6] = "Positive-sequence Reactance, ohms per unit length. Setting any of R1, R0, X1, X0, C1, C0 forces " +
				"the program to use the symmetrical component line definition.  See also Xmatrix";
		propertyHelp[7] = "Zero-sequence Resistance, ohms per unit length.";
		propertyHelp[8] = "Zero-sequence Reactance, ohms per unit length.";
		propertyHelp[9] = "Positive-sequence capacitance, nf per unit length.  Setting any of R1, R0, X1, X0, C1, C0 forces " +
				"the program to use the symmetrical component line definition. See also Cmatrix.";
		propertyHelp[10] = "Zero-sequence capacitance, nf per unit length.";
		propertyHelp[11] = "Resistance matrix, lower triangle, ohms per unit length. Order of the matrix is the number of phases. "+
				"May be used to specify the impedance of any line configuration. Using any of Rmatrix, Xmatrix, Cmatrix " +
				"forces program to use the matrix values for line impedance definition. For balanced line models, you may "+
				"use the standard symmetrical component data definition instead.";
		propertyHelp[12] = "Reactance matrix, lower triangle, ohms per unit length. Order of the matrix is the number of phases. "+
				"May be used to specify the impedance of any line configuration. Using any of Rmatrix, Xmatrix, Cmatrix " +
				"forces program to use the matrix values for line impedance definition.  For balanced line models, you may "+
				"use the standard symmetrical component data definition instead.";
		propertyHelp[13] = "Nodal Capacitance matrix, lower triangle, nf per unit length.Order of the matrix is the number of phases. "+
				"May be used to specify the shunt capacitance of any line configuration. Using any of Rmatrix, Xmatrix, Cmatrix " +
				"forces program to use the matrix values for line impedance definition.  For balanced line models, you may "+
				"use the standard symmetrical component data definition instead.";
		propertyHelp[14] = "{y/n | T/F}  Default= no/false.  Designates this line as a switch for graphics and algorithmic purposes. " +DSS.CRLF+
				"SIDE EFFECT: Sets r1 = 1.0; x1 = 1.0; r0 = 1.0; x0 = 1.0; c1 = 1.1 ; c0 = 1.0;  length = 0.001; You must reset if you want something different.";
		propertyHelp[15] = "Carson earth return resistance per unit length used to compute impedance values at base frequency. " +
				"Default is 0.01805 = 60 Hz value in ohms per kft (matches default line impedances). " +
				"This value is required for harmonic solutions if you wish to adjust the earth return impedances for frequency. " +
				"If not, set both Rg and Xg = 0.";
		propertyHelp[16] = "Carson earth return reactance per unit length used to compute impedance values at base frequency.  For making better frequency adjustments. " +
				"Default is 0.155081 = 60 Hz value in ohms per kft (matches default line impedances). " +
				"This value is required for harmonic solutions if you wish to adjust the earth return impedances for frequency. " +
				"If not, set both Rg and Xg = 0.";
		propertyHelp[17] = "Default=100 meter ohms.  Earth resitivity used to compute earth correction factor. Overrides Line geometry definition if specified.";
		propertyHelp[18] = "Geometry code for LineGeometry Object. Supercedes any previous definition of line impedance. " +
				"Line constants are computed for each frequency change or rho change. CAUTION: may alter number of phases. "+
				"You cannot subsequently change the number of phases unless you change how the line impedance is defined.";
		propertyHelp[19] = "Length Units = {none | mi|kft|km|m|Ft|in|cm } Default is None - assumes length units match impedance units.";
		propertyHelp[20] = "Reference to a LineSpacing for use in a line constants calculation." + DSS.CRLF +
				"Must be used in conjunction with the Wires property." + DSS.CRLF +
				"Specify this before the wires property.";
		propertyHelp[21] = "Array of WireData names for use in a line constants calculation." + DSS.CRLF +
				"Must be used in conjunction with the Spacing property." + DSS.CRLF +
				"Specify the Spacing first, and \"ncond\" wires." + DSS.CRLF +
				"May also be used to specify bare neutrals with cables, using \"ncond-nphase\" wires.";
		propertyHelp[22] = "One of {Carson | FullCarson | Deri}. Default is the global value established with the Set EarthModel command. " +
				"See the Options Help on EarthModel option. This is used to override the global value for this line. This " +
				"option applies only when the \"geometry\" property is used.";
		propertyHelp[23] = "Array of CNData names for use in a cable constants calculation." + DSS.CRLF +
				"Must be used in conjunction with the Spacing property." + DSS.CRLF +
				"Specify the Spacing first, using \"nphases\" cncables." + DSS.CRLF +
				"You may later specify \"nconds-nphases\" wires for separate neutrals";
		propertyHelp[24] = "Array of TSData names for use in a cable constants calculation." + DSS.CRLF +
				"Must be used in conjunction with the Spacing property." + DSS.CRLF +
				"Specify the Spacing first, using \"nphases\" tscables." + DSS.CRLF +
				"You may later specify \"nconds-nphases\" wires for separate neutrals";

		activeProperty = Line.NumPropsThisClass - 1;
		super.defineProperties();  // add defs of inherited properties to bottom of list
	}

	@Override
	public int newObject(String objName) {
		DSS.activeCircuit.setActiveCktElement(new LineObj(this, objName));
		return addObjectToList(DSS.activeDSSObject);
	}

	private void doRmatrix() {
		int[] nOrder = new int[1];
		Complex[] Zvalues;

		LineObj elem = activeLineObj;

		double[] matBuffer = new double[elem.getNumPhases() * elem.getNumPhases()];
		int orderFound = Parser.getInstance().parseAsSymMatrix(elem.getNumPhases(), matBuffer);

		if (orderFound > 0) {  // parse was successful
			Zvalues = elem.getZ().asArray(nOrder);
			if (nOrder[0] == elem.getNumPhases()) {
				for (int j = 0; j < elem.getNumPhases() * elem.getNumPhases(); j++)
					Zvalues[j] = new Complex(matBuffer[j], Zvalues[j].getImaginary());
			}
		}
	}

	private void doXmatrix() {
		int[] nOrder = new int[1];
		Complex[] Zvalues;

		LineObj elem = activeLineObj;

		double[] matBuffer = new double[elem.getNumPhases() * elem.getNumPhases()];
		int orderFound = Parser.getInstance().parseAsSymMatrix(elem.getNumPhases(), matBuffer);

		if (orderFound > 0) {  // parse was successful
			Zvalues = elem.getZ().asArray(nOrder);
			if (nOrder[0] == elem.getNumPhases()) {
				for (int j = 0; j < elem.getNumPhases() * elem.getNumPhases(); j++)
					Zvalues[j] = new Complex(Zvalues[j].getReal(), matBuffer[j]);
			}
		}
	}

	private void doCmatrix() {
		int[] nOrder = new int[1];
		Complex[] Yvalues;
		double factor;

		LineObj elem = activeLineObj;

		double[] matBuffer = new double[elem.getNumPhases() * elem.getNumPhases()];
		int orderFound = Parser.getInstance().parseAsSymMatrix(elem.getNumPhases(), matBuffer);

		if (orderFound > 0) {  // parse was successful
			factor = DSS.TWO_PI * elem.getBaseFrequency() * 1.0e-9;
			Yvalues = elem.getYc().asArray(nOrder);
			if (nOrder[0] == elem.getNumPhases()) {
				for (int j = 0; j < elem.getNumPhases() * elem.getNumPhases(); j++)
					Yvalues[j] = new Complex(Yvalues[j].getReal(), factor * matBuffer[j]);
			}
		}
	}

	/**
	 * A line defaults to 3-phases and some typical symmetrical component data.
	 *
	 * Line impedances are specified in per unit length and are multiplied by the length
	 * when the primitive Y matrix is computed.
	 *
	 * You may specify the impedances of the line either by symmetrical components or
	 * by R, X, and nodal C matrices  (also per unit length).
	 *
	 * All C's is entered in nano farads.
	 *
	 * The ultimate values are in the matrices. If you specify matrices, then the symmetrical
	 * component values are ignored. However, if you change any of the symmetrical component values
	 * the matrices will be recomputed. It is assumed you want to use symmetrical component values.
	 * Don't mix data entry by matrix and by symmetrical components.
	 *
	 * Note that if you change the number of phases, the matrices are reallocated and reinitialized
	 * with whatever is currently in the symmetrical component data.
	 */
	@Override
	public int edit() {
		Parser parser = Parser.getInstance();

		LineUnits newLengthUnits;

		// continue parsing with contents of parser
		activeLineObj = (LineObj) elementList.getActive();
		DSS.activeCircuit.setActiveCktElement(activeLineObj);  // use property to set this value

		LineObj elem = activeLineObj;

		int paramPointer = -1;
		String paramName = parser.getNextParam();
		String param = parser.stringValue();

		while (param.length() > 0) {
			if (paramName.length() == 0) {
				paramPointer += 1;
			} else {
				paramPointer = commandList.getCommand(paramName);
			}

			if (paramPointer >= 0 && paramPointer < getNumProperties())
				elem.setPropertyValue(paramPointer, param);

			switch (paramPointer) {
			case -1:
				DSS.doSimpleMsg("Unknown parameter \"" + paramName + "\" for object \"Line." +
						elem.getName() + "\"", 181);
				break;
			case 0:
				elem.setBus(0, param);
				break;
			case 1:
				elem.setBus(1, param);
				break;
			case 2:
				elem.fetchLineCode(param);  // define line by conductor code
				break;
			case 3:
				elem.setLen(parser.doubleValue());
				break;
			case 4:
				/* nPhases (see below) */
				break;
			case 5:
				elem.setR1(parser.doubleValue());
				break;
			case 6:
				elem.setX1(parser.doubleValue());
				break;
			case 7:
				elem.setR0(parser.doubleValue());
				break;
			case 8:
				elem.setX0(parser.doubleValue());
				break;
			case 9:
				elem.setC1(parser.doubleValue() * 1.0e-9);  // convert from nano to farads
				break;
			case 10:
				elem.setC0(parser.doubleValue() * 1.0e-9);
				break;
			case 11:
				doRmatrix();
				break;
			case 12:
				doXmatrix();
				break;
			case 13:
				doCmatrix();
				break;
			case 14:
				elem.setSwitch(interpretYesNo(param));
				break;
			case 15:
				elem.setRg(parser.doubleValue());
				break;
			case 16:
				elem.setXg(parser.doubleValue());
				break;
			case 17:
				elem.setRho(parser.doubleValue());
				elem.setRhoSpecified(true);
				break;
			case 18:
				elem.fetchGeometryCode(param);
				break;
			case 19:  // update units conversion factor that might have been changed previously
				newLengthUnits = LineUnits.interpretUnitsCode(param);
				if (elem.isLineCodeSpecified()) {
					elem.setUnitsConvert(LineUnits.convertLineUnits(elem.getLineCodeUnits(), newLengthUnits));
				} else {
					elem.setUnitsConvert(elem.getUnitsConvert() * LineUnits.convertLineUnits(elem.getLengthUnits(), newLengthUnits));
				}
				elem.setLengthUnits(newLengthUnits);
				break;
			case 20:
				elem.fetchLineSpacing(param);
				break;
			case 21:
				elem.fetchWireList(param);
				break;
			case 22:
				elem.setEarthModel(interpretEarthModel(param));
				break;
			case 23:
				elem.fetchCNCableList(param);
				break;
			case 24:
				elem.fetchTSCableList(param);
				break;
			default:
				// inherited property edits
				classEdit(activeLineObj, paramPointer - Line.NumPropsThisClass);
				break;
			}

			// side effects ...
			switch (paramPointer) {
			case 2:
				elem.setSpacingSpecified(false);
				if (elem.isGeometrySpecified() == true)
					elem.killGeometrySpecified();
				elem.setGeometrySpecified(false);
				break;
			case 4:  /* Change the number of phases ... only valid if symComponentsModel=true */
				if (elem.getNumPhases() != parser.integerValue())
					if (!elem.isGeometrySpecified() && elem.isSymComponentsModel()) {  // ignore change of nPhases if geometry used
						elem.setNumPhases(parser.integerValue());
						elem.setNumConds(elem.getNumPhases());  // force reallocation of terminal info
						elem.setYOrder(elem.getNumTerms() * elem.getNumConds());
						/*al.setYprimInvalid(true);*/  // now set below
						elem.recalcElementData();  // reallocate Z, etc.
					} else {
						DSS.doSimpleMsg("Illegal change of number of phases for Line."+elem.getName(), 181);
					}
				break;
			case 5:
			case 6:
			case 7:
			case 8:
			case 9:
			case 10:
				elem.setLineCodeSpecified(false);
				elem.killGeometrySpecified();
				elem.killSpacingSpecified();
				elem.resetLengthUnits();
				elem.setSymComponentsChanged(true);
				elem.setSymComponentsModel(true);
				break;
			case 11:
			case 12:
			case 13:
				elem.setLineCodeSpecified(false);
				elem.setSymComponentsModel(false);
				elem.resetLengthUnits();
				elem.killGeometrySpecified();
				elem.killSpacingSpecified();
				break;
			case 14:
				if (elem.isSwitch()) {
					elem.setSymComponentsChanged(true);
					elem.setYPrimInvalid(true);
					elem.setGeometrySpecified(false);
					elem.setSpacingSpecified(false);
					elem.setR1(1.0);
					elem.setX1(1.0);
					elem.setR0(1.0);
					elem.setX0(1.0);
					elem.setC1(1.1 * 1.0e-9);
					elem.setC0(1.0 * 1.0e-9);
					elem.setLen(0.001);
					elem.resetLengthUnits();
				}
				break;
			case 16:
			case 17:
				elem.setKXg(elem.getXg() / log(658.5 * sqrt(elem.getRho() / elem.getBaseFrequency())));
				break;
			case 18:
				elem.setGeometrySpecified(true);
				elem.setSymComponentsModel(false);
				elem.setSymComponentsChanged(false);
				break;
			case 20:
			case 21:
			case 23:
			case 24:
				if (elem.getLineSpacingObj() != null && elem.getWireData() != null) {
					elem.setSpacingSpecified(true);
					elem.setSymComponentsModel(false);
					elem.setSymComponentsChanged(false);
					elem.killGeometrySpecified();
				}
				elem.setYPrimInvalid(true);
				break;
			}

			// YPrim invalidation on anything that changes impedance values
			if (paramPointer >= 2 && paramPointer <= 13) {
				elem.setYPrimInvalid(true);
			} else if (paramPointer == 17) {
				if (elem.isGeometrySpecified() && (elem.getLineGeometryObj() != null))
					elem.getLineGeometryObj().setRhoEarth(elem.getRho());
			}

			switch (paramPointer) {
			case 9:
			case 10:
			case 13:
				elem.setCapSpecified(true);
				break;
			}

			paramName = parser.getNextParam();
			param = parser.stringValue();
		}

		//if (elem.isSymComponentsChanged()) elem.recalcElementData();
		return 0;
	}

	@Override
	protected int makeLike(String lineName) {
		int success = 0;

		/* See if we can find this line name in the present collection */
		LineObj other = (LineObj) find(lineName);

		if (other != null) {
			LineObj elem = activeLineObj;

			if (elem.getNumPhases() != other.getNumPhases()) {
				elem.setNumPhases(other.getNumPhases());
				elem.setNumConds(elem.getNumPhases());  // force reallocation of terminals and conductors

				elem.setYOrder(elem.getNumConds() * elem.getNumTerms());
				elem.setYPrimInvalid(true);

				// for a line, nPhases = nCond, for now
				elem.setZ(new CMatrix(elem.getNumPhases()));
				elem.setZinv(new CMatrix(elem.getNumPhases()));
				elem.setYc(new CMatrix(elem.getNumPhases()));
			}

			elem.getZ().copyFrom(other.getZ());
			// al.getZinv().copyFrom(OtherLine.getZinv());
			elem.getYc().copyFrom(other.getYc());
			elem.setR1(other.getR1());
			elem.setX1(other.getX1());
			elem.setR0(other.getR0());
			elem.setX0(other.getX0());
			elem.setC1(other.getC1());
			elem.setC0(other.getC0());
			elem.setLen(other.getLen());
			elem.setSymComponentsModel(other.isSymComponentsModel());
			elem.setCapSpecified(other.isCapSpecified());

			classMakeLike(other);  // take care of inherited class properties

			for (int i = 0; i < elem.getParentClass().getNumProperties(); i++)
				elem.setPropertyValue(i, other.getPropertyValue(i));

			success = 1;
		} else {
			DSS.doSimpleMsg("Error in Line makeLike: \"" + lineName + "\" not found.", 182);
		}

		return success;
	}

	@Override
	public int init(int handle) {
		DSS.doSimpleMsg("Need to implement Line.init()", -1);
		return 0;
	}

}
